### PR TITLE
90 Implement curriculum learning

### PR DESCRIPTION
**Pull request by GustawB**
This PR introduces an algorithm to first balance classes distribution in the learning data, and the gradually skew it during training. Additionally, this algorithm is added to the training pipeline.

27.02.25 UPDATE:
1. Add accuracy metric for the whole result AND for each class separately.
2. Check if only classification head is trained instead of everything after embedding. I will then consult whether possibly fixing this issue is not too much for my task.

### CODE REVIEW

**Comment by GustawB** [comment]
About training the whole model:
It seems that the Trainer class is a "simple" pytorch wrapper; on training, it just calls .train() on model (in function train_step: https://github.com/huggingface/transformers/blob/main/src/transformers/trainer.py#L3701). Because models are pytorch modules, my thesis is that all weights are being set to the train, not only those of the classification head. It also seems that I'm not alone in this conclusion; I found multiple forum posts that had the same assumption (https://discuss.huggingface.co/t/how-to-freeze-layers-using-trainer/4702/2; https://discuss.huggingface.co/t/the-point-of-using-pretrained-model-if-i-dont-freeze-layers/40675) and some reddit stuff (https://www.reddit.com/r/deeplearning/comments/ndmqm6/why_do_we_train_whole_bert_model_for_fine_tuning/). So, it seems that the Trainer class, by default, trains the whole model, not just the classification head.

File: src/bert_finetuning/README.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
+# Curriculum learning for NER
+During my experiments with BERT I noticed that our data has a lot of noise; coupon-to-non-coupon ratio was around 6%.
+In order to balance classes, but still enable the model to retrieve coupons from the noisy data, I propose this algorithm:
+1. Split rows that have coupons and rows that don't have coupons.
+2. Represent every row by spans; initially, each span should contain coupons and their neighbourhood, so that spans in each row contain 50/50 split (if possible).
```
Comment (kdybek) [inline src/bert_finetuning/README.md:5]: I don't get what are spans. Maybe it's just me, but this description seems a little hazy

File: src/bert_finetuning/README.md [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+3. Create n datasets (n specified by the user) in the following manner:
+    1. If current dataset has e.g. index, add x/n rows with no coupons to the dataset (x ins the init number of rows without coupons).
```
Comment (kdybek) [inline src/bert_finetuning/README.md:7]: This line is also vague. What does it mean that a dataset has e.g. index?

**Review by kdybek — CHANGES_REQUESTED** [review]
Looks good, I think. Apart from the comments that I left, I think there should be a section in the readme regarding the dataset changing policy. Do you use one dataset for a fixed number of epochs before switching to the next one, or maybe you switch after reaching some accuracy or loss threshold?

**Review by SzymonKozl — CHANGES_REQUESTED** [review]
in overall gj, leaving some small requests and questions, mostly related to the documentation.

File: src/bert_finetuning/curriculer.py [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+    def __eq__(self, other_id: int):
+        return self.row_id == other_id
+
+    def __lt__(self, other_id: int):
+        return self.row_id < other_id
+
+
+class SpanData:
+    def __init__(self):
+        self.beg:int  = -1
```
[the remark this answers is not in the document]
Comment (GustawB) [reply src/bert_finetuning/curriculer.py:27]: done

File: src/bert_finetuning/finetuner.py [hunk 4]
```diff
@@ -169,26 +170,43 @@ def __compute_metrics(custom_labels: list, eval_preds: list) -> dict:
...
+    result = {
         "precision": all_metrics["overall_precision"],
         "recall": all_metrics["overall_recall"],
         "f1": all_metrics["overall_f1"],
-        "accuracy": all_metrics["overall_accuracy"],
+        "overall_accuracy": all_metrics["overall_accuracy"],
     }
 
+    for label in per_class:
+        result[label + "_accuracy"] = per_class[label][0] / per_class[label][1]
```
[the remark this answers is not in the document]
Comment (GustawB) [reply src/bert_finetuning/finetuner.py:191]: done

File: test/test_curriculer.py [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+        self.init_c = {
+             0: [(1, 2, 2), (5,5, 1), (7, 8, 2), (10, 12, 3)],
+             3: [(1, 4, 4), (5, 5, 1), (7, 8, 2), (10, 11, 2)],
+             4: [(0, 5, 6)],
+             6: [(1, 2, 2), (10, 11, 2)],
+             8: [(2, 3, 2)],
+        }
+
+        self.init_non_c = [1, 2, 5, 7]
+
```
[the remark this answers is not in the document]
Comment (GustawB) [reply test/test_curriculer.py:193]: done

File: test/test_curriculer.py [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+
+
+        yield
+
```
[the remark this answers is not in the document]
Comment (GustawB) [reply test/test_curriculer.py:197]: Done

File: src/bert_finetuning/curriculer.py [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+            rid = int(row.row_id)
+            col = 'train'
+            if row.row_id >= train_len + validation_len:
+                rid -= train_len + validation_len
+                col = 'test'
+            elif row.row_id >= train_len:
+                rid -= train_len
+                col = 'validation'
+            for span in row.spans:
+                labels_list.extend(self.__dataset[col][rid]['labels'][int(span.beg):int(span.end) + 1])
```
[the remark this answers is not in the document]
Comment (GustawB) [reply src/bert_finetuning/curriculer.py:120]: I was dividing something earlier; it was being casted to float; I think I fixed it later (it was earlier in the code) but I didn't think it was necessary to remove it here (ugabuga nice constraint).

[the remark this answers is not in the document]
Comment (GustawB) [reply src/bert_finetuning/curriculer.py:120]: nvm deleted

File: src/bert_finetuning/README.md (hunk 2 above)
Comment (GustawB) [reply src/bert_finetuning/README.md:7]: done

File: src/bert_finetuning/curriculer.py [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+
+    def length(self):
+        return self.end - self.beg + 1
+
+
+class Curriculer:
+    def __init__(self, dataset: Dataset, splits_amount: int):
+        """
+        :param dataset: Dataset object with 'train', 'validation' and 'test' keys
+        :param splits_amount: Amount of splits to perform
```
[the remark this answers is not in the document]
Comment (GustawB) [reply src/bert_finetuning/curriculer.py:38]: done

File: src/bert_finetuning/curriculer.py [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+            total_l = 0
+            for i in range(len(row.spans)):
+                total_l += row.spans[i].length()
+            extend_spans(row.spans, total_l, row.max_size)
+        self.__init_len = len(self.__rows_with_c)
+        return self.__create_dataset(tv_split, tt_split)
+
+    def yield_dataset(self, tv_split: float = 0.2, tt_split: float = 0.5, shuffle: bool = True) -> DatasetDict:
+        """
+        Main generation logic. Depending on whether the current iteration is even or odd,
```
[the remark this answers is not in the document]
Comment (GustawB) [reply src/bert_finetuning/curriculer.py:175]: done

File: src/bert_finetuning/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (SzymonKozl) [reply src/bert_finetuning/README.md]: what is n here?

[the remark this answers is not in the document]
Comment (GustawB) [reply src/bert_finetuning/README.md]: N datasets

[the remark this answers is not in the document]
Comment (GustawB) [reply src/bert_finetuning/README.md]: Like 5 datasets

File: src/bert_finetuning/curriculer.py [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+    """
+    # Find the index where the target_value *should* be
+    index = bisect_left(rows, target_value)
+
+    # Check if the index is valid and actually matches the target value
+    if index < len(rows) and rows[index].row_id == target_value:
+        return index  # Return the found object's index
+    return None  # Not found
+
+def extend_spans(spans: list, extend_amount: int, max_len: int):
```
[the remark this answers is not in the document]
Comment (GustawB) [reply src/bert_finetuning/curriculer.py:221]: done; fixed bug, added asserts to test

File: src/bert_finetuning/curriculer.py [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+    while extend_amount > 0 and k < spans_count and spans_len < max_len:
+        beg = spans[k].beg
+        end = spans[k].end
+        if k == 0:
+            spans[k].beg = max(0, spans[k].beg - extend_amount)
+        else:
+            spans[k].beg = max(spans[k - 1].end + 1, spans[k].beg - extend_amount)
+        extend_amount -= beg - spans[k].beg
+        if k == spans_count - 1:
+            spans[k].end = min(max_len - 1, spans[k].end + extend_amount)
```
[the remark this answers is not in the document]
Comment (GustawB) [reply src/bert_finetuning/curriculer.py:267]: It is possible; however, notice that I do two passes over the row; in the first one, I try to balance the distribution of zeros added; then it is possible that I cant extend the last span. but then, the line that you mentioned is in the second loop, that greedily adds zeros wherever it can; so, if we came to the last span, and we can extend it so that extend_amount ==0, it means that spans already cover the whole row.

File: src/bert_finetuning/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (GustawB) [reply src/bert_finetuning/README.md]: still though, this part is no longer in the readme.

**Review by kdybek — APPROVED** [review]
LGTM

File: src/bert_finetuning/curriculer.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (SzymonKozl) [reply src/bert_finetuning/curriculer.py:265]: ok fair

**Review by SzymonKozl — APPROVED** [review]

**Review by kdybek — CHANGES_REQUESTED** [review]
I think there is a bug in __align_labels_with_tokens function. If a B-X word gets split into a couple tokens, every token will be B-X.

File: src/bert_finetuning/modal_script.py [hunk 12]
```diff
@@ -41,7 +42,7 @@ def run_fine_tuning(hf_token, wandb_key, dataset_name, push_to_hub=False):
         label2id=label2id,
     )
 
-    ft.train_model(model, tokenized_dataset, labels, wandb_log=True, run_name=dataset_name)
+    ft.train_model(model, cs, labels, wandb_log=True, run_name=dataset_name, curriculum_learning=True)
```
Comment (kdybek) [inline src/bert_finetuning/modal_script.py:45]: Why did you change tokenized_dataset to cs here?

File: src/bert_finetuning/modal_script.py [hunk 13]
```diff
@@ -23,7 +24,7 @@
 )
```
Comment (kdybek) [inline src/bert_finetuning/modal_script.py:24]: .env({"HALT_AND_CATCH_FIRE": 0}) does not work for me because 0 is not a string. Did it work in your case?

File: src/bert_finetuning/modal_script.py (hunk 12 above)
Comment (GustawB) [reply src/bert_finetuning/modal_script.py:45]: image

Because curriculer performs its own tokenizing and aligning; fair point though that then the default approach should also do this.

File: src/bert_finetuning/modal_script.py (hunk 13 above)
Comment (GustawB) [reply src/bert_finetuning/modal_script.py:24]: Can you post the error? It works for me, I took it from docs

Comment (GustawB) [reply src/bert_finetuning/modal_script.py:24]: and how do you run the script?

File: src/bert_finetuning/modal_script.py (hunk 12 above)
Comment (GustawB) [reply src/bert_finetuning/modal_script.py]: fixed; you can find the fix in the finetuner (moved tokenizing there)

File: src/bert_finetuning/modal_script.py (hunk 13 above)
Comment (kdybek) [reply src/bert_finetuning/modal_script.py:24]: I used your bash script. It's an easy fix anyway so let's leave it as is ig

File: src/bert_finetuning/modal_script.py (hunk 12 above)
Comment (kdybek) [reply src/bert_finetuning/modal_script.py]: You are tokenizing the dataset two times after the 833a14e fix.

**Comment by GustawB** [comment]
> I think there is a bug in __align_labels_with_tokens function. If a B-X word gets split into a couple tokens, every token will be B-X.

This was intentional, however after clarification I agree that this is not a proper behaviour. I fixed this issue.

File: src/bert_finetuning/modal_script.py (hunk 13 above)
Comment (GustawB) [reply src/bert_finetuning/modal_script.py:24]: still, for reference; HALT_AND_CATCH expects string; I gave it an int (from docs) and it was working, but it seems that it silently swapped 0 for an empty string for me, while for Kamil it defaulted to an error. Fixed.

**Review by kdybek — APPROVED** [review]
LGTM fr no cap this time
